gax/src/test/java/com/google/api/gax/rpc/ServerStreamTest.java

```diff
@@ -0,0 +1,201 @@
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.SettableApiFuture;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Queues;
+import com.google.common.truth.Truth;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ServerStreamTest {
+  private TestStreamController controller;
+  private ServerStream<Integer> stream;
+  private ExecutorService executor;
+
+  @Rule public ExpectedException expectedException = ExpectedException.none();
+
+  @Before
+  public void setUp() throws Exception {
+    controller = new TestStreamController();
+    stream = new ServerStream<>();
+
+    stream.observer().onStart(controller);
+    executor = Executors.newSingleThreadExecutor();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    executor.shutdownNow();
+  }
+
+  @Test
+  public void testEmptyStream() {
+    stream.observer().onStart(controller);
+    stream.observer().onComplete();
+
+    Truth.assertThat(Lists.newArrayList(stream)).isEmpty();
+  }
+
+  @Test
+  public void testMultipleItemStream() throws InterruptedException {
+    executor.submit(
+        new Callable<Void>() {
+          @Override
+          public Void call() throws InterruptedException {
+            for (int i = 0; i < 5; i++) {
+              int requestCount = controller.requests.poll(1, TimeUnit.SECONDS);
+
+              Truth.assertWithMessage("ServerStream should request one item at a time")
+                  .that(requestCount)
+                  .isEqualTo(1);
+
+              stream.observer().onResponse(i);
+            }
+            stream.observer().onComplete();
+            return null;
+          }
+        });
+
+    Truth.assertThat(Lists.newArrayList(stream)).containsExactly(0, 1, 2, 3, 4);
+  }
+
+  @Test
+  public void testEarlyTermination() throws Exception {
+    executor.submit(
+        new Callable<Void>() {
+          @Override
+          public Void call() throws InterruptedException, ExecutionException, TimeoutException {
+            int i = 0;
+            while (controller.requests.poll(500, TimeUnit.MILLISECONDS) != null) {
+              stream.observer().onResponse(i++);
+            }
+            Throwable cancelException = controller.cancelFuture.get(1, TimeUnit.SECONDS);
+            stream.observer().onError(cancelException);
+            return null;
```

Comment:
Same comment about this not actually failing the test.

Reply:
fixed